lib/matplotlib/table.py

```diff
@@ -674,7 +676,7 @@ def table(ax,
 
     Parameters
     ----------
-    cellText : 2D list of str, optional
+    cellText : 2D list of str, Pandas.DataFrame, optional
```

Comment:
I'm not sure we want to call out pandas data frames here.  How do we do this elsewhere?  I'd just state "object" and then below, state that 
"""
If `cellText` is not a list of texts, we attempt to access `cellText.columns.to_numpy()` for the column headers, 
and `cellText.to_numpy()` for the table's cells.
"""

Reply:
We don't have any other functions explicitly take DataFrames. Typically, DataFrame is covered by what we call `array-like` (and internally runs via `to_numpy()`. It's different here because we take the column headers as well. Due to this particular usage, we've chosen to explicitly check for DataFrames (`_is_pandas_dataframe()`). I believe that duck-typing would be a bit too vague here: An object with `obj.columns.to_numpy()` and `obj.to_numpy()` is very particular. I'm not sure if this matches anything but pandas.DataFrame - and if there exists other such objects, I'm not clear that accepting them is appropriate. If the need for similar objects should arise, we can always revisit. For now, the explicit type is the simplest documentation and matches implementation:

```suggestion
    cellText : 2D list of str or pandas.DataFrame, optional
```

Reply:
I linked an example of another very popular package, polars, that has the same form of data frame.  I'm sure there are others out there.  I don't think we should be special casing pandas here.

Reply:
Polars is logically equivalent, but it does not have the same API. See https://github.com/matplotlib/matplotlib/pull/28830#discussion_r1773029308.

There are several other dataframes out their (https://data-apis.org/dataframe-api/draft/purpose_and_scope.html#history-and-dataframe-implementations), but the problem is that their API is currently not consistent. IMHO we therefore would have to special case to support multiple of them.

While I appreciate the attempt to duck-type instead of explicitly relying on types, I believe the current dataframe APIs are to inhomogeneous to support reasonable duck-typing. Any object that supports `obj.columns.to_numpy()` and `obj.to_numpy()` is too technical - i suspect that at least half of the pandas users would not know whether that holds for `pandas.DataFrame` or not. Duck-typing would work if `dataframe-like` was well-defined, but it is currently not (see the draft dataframe API standard).

Reply:
I think `np.asanyarray(df.columns)` works on both pandas and polars, and is pretty straightforward.

Reply:
If you really want to push for duck-typing, go for it. But I'd require reasonable type stubs (accepting `Any` is not good enough) and type documentation (users should at least easily see that pandas.DataFrame as the most common dataframe type is supported).

Ideally, that claimed behavior would also be tested. (It's easy for pandas, because it's a test dependency. For others, you'd likely need a mock similar to `test_unpack_to_numpy_from_torch`.)

I personally believe it's not worth the added effort and would advise against requiring this from a first-time contribution. When accepting the limited scope of pandas, this PR is almost ready apart from minor documentation issues and the type stub. I therefore suggest to take this PR as a limited but well scoped improvement. We don't paint us in any corner with that. You can generalize in a followup PR.

Reply:
> But I'd require reasonable type stubs 

OK, but I dont' think you will get pandas dataframes in a typestub from this PR either, will you?  I'm personally pretty :-1: on losing functionality because of type stubbing issues.